dnsutil.py

```diff
@@ -169,67 +174,139 @@ def resolvconf_update(setup: bool = True, searches: Optional[list[str]] = None)
+                result.append(line)
+    return result
+
+
+def _dnsmasq_servers_content_update(old_content: bytes, cluster_name: Optional[str], api_vip: Optional[str] = None) -> tuple[bytes, list[bytes]]:
+
+    old_entries = _dnsmasq_servers_content_parse(old_content)
+
+    new_entries = list(old_entries)
+
+    if cluster_name is None:
         pass
+    elif api_vip is None:
+        for entry in new_entries:
+            if entry.startswith(f"server=/{cluster_name}.redhat.com/".encode()):
+                new_entries.remove(entry)
     else:
-        resolvconf_update(searches=rcdata.searches)
-
-    with common.atomic_write(dmasqconf_orig) as file:
-        content = "# Written by cluster-deployment-automation.\n"
-        content += "# Contains the nameserves from \"/etc/resolv.conf.cda-orig\".\n"
-        content += ''.join(f"server={n}\n" for n in rcdata.nameservers)
-        content += "listen-address=127.0.0.1\n"
-        content += "bind-interfaces\n"
-        file.write(content)
-
-    if service_restart:
-        ret = lh.run("systemctl unmask dnsmasq.service")
-        ret = lh.run("systemctl restart dnsmasq.service")
-        if not ret.success():
-            logger.warning(f"dnsmasq: failure to start dnsmasq service: {ret}")
-        else:
-            logger.debug("dnsmasq: service was restarted")
-    else:
-        logger.debug("dnsmasq: service was not restarted")
+        new_entry = f"server=/{cluster_name}.redhat.com/{api_vip}".encode()
+        if new_entry not in new_entries:
+            new_entries.append(new_entry)
+
+    new_entries.sort()
+
+    new_content = (
+        b"""# Written by cluster-deployment-automation for resolving cluster names.
+# This file is passed to dnsmasq via the --servers-file= option
+#
+# Reload after changes with
+#   systemctl restart dnsmasq.service
+#   pkill -HUP -f '\\<dnsmasq\\>.*conf-file=/var/lib/libvirt/dnsmasq/default.conf'
+"""
+        + b"\n".join(new_entries)
+        + b"\n"
+    )
+
+    return new_content, new_entries
+
+
+def _dnsmasq_servers_update(cluster_name: Optional[str], api_vip: Optional[str] = None) -> bool:
+
+    try:
+        with open(DNSMASQ_SERVERS_FILE, "rb") as f:
+            old_content = f.read()
+    except Exception:
+        old_content = b""
+
+    new_content, new_entries = _dnsmasq_servers_content_update(old_content, cluster_name, api_vip)
+
+    if new_content == old_content:
+        return False
+
+    logger.debug(f"dnsmasq: update {DNSMASQ_SERVERS_FILE} (content: {[s.decode('utf-8', errors='replace') for s in new_entries]})")
+
+    os.makedirs(os.path.dirname(DNSMASQ_SERVERS_FILE), exist_ok=True)
+
+    with common.atomic_write(DNSMASQ_SERVERS_FILE, text=False) as f:
+        f.write(new_content)
+
+    return True
+
+
+def dnsmasq_servers_update(cluster_name: Optional[str], api_vip: Optional[str] = None) -> bool:
+    with _lock:
+        return _dnsmasq_servers_update(cluster_name, api_vip)
+
+
+def dnsmasq_servers_reload() -> None:
```

Comment:
pass in host?

Reply:
most of the API in `dnsutil` is not host-agnostic, because it directly writes files.

There is also no point in trying to send every host-specific operation through a `Host.run()`, because this functionality is explicitly only useful for the provisioning host (localhost).

At that point, `dnsmasq_servers_reload()` also only makes sense at localhost and it doesn't need an argument that can reasonably only be `host.LocalHost()`.

Reply:
Shouldn't it use `host.write(...)` ?

Reply:
Also, is the provisioning host always `localhost`?

Reply:
> Also, is the provisioning host always localhost?

I mean, the host where you type `./cda.py` and where the python code runs. It's also the one that creates the virbr0 interfaces and connects the 192.168.122.x network with local VMs and baremetal. It's the one that gets a local dnsmasq.service running.

I call that  "the provisioning host" and that's localhost from point of view of the running python code.

Reply:
yes

Reply:
I don't think so. We pass `lh = host.LocalHost()` in multiple places like. Are you saying that we should change those calls?

Reply:
I am talking specifically about the API in `dnsutil`. That API is only used/useful on the provisioning host (ie. localhost). Which is why is has no `Host` parameter to operate on another host.

I don't understand what you mean.

Reply:
To clarify, other places in CDA don't assume that `provisioning host == localhost`. Why would we want to make that assumption here?

Reply:
I don't know what you mean by "provisioning host".

I said, in my vocabulary, the "provisioning host" is always localhost. It's the host that runs the CDA code. If you have another meaning of that word, then it would be good to clarify what you mean.

Beyond that. Setting up the local dnsmasq service is *only* done on the "provisioning host" (or whatever we call this). It never happens on another host, it's not planned to ever do it, and it probably(*) makes no sense.

(*) if you disagree with that, you have to explain how it would make sense.

> Why would we want to make that assumption here?

Because it matters what those places do. Some places may do something on remote hosts. `dnsutil` does not.

Reply:
Let us know if it works for you. If we have one lgtm, I will merge it.

Reply:
Imagine 500 remote VMs accessing a single drive on the provision host.

Reply:
That isn't relevant because this isn't about `NFS` and because current CDA also doesn't handle `NFS` the way you seem to suggest.

I said, that there is no use-case for remoting what `dnsutil`. No other hosts in the cluster should configure a similar dnsmasq.service. They should use the one of the provisioning host. Which they do with this patch (via `virbr0`).

Before requesting to implement something, you should at least argue that there is (at least a somewhat plausible) use case for it. Beyond "imagine I run this on a remote host and then I setup the dnsmasq.service on that host", but how that is ultimately useful beyond the surface level.

Reply:
ok, we can change it later if we need to.